source/ui/Components/editor/Editor.qml

```diff
@@ -170,102 +199,488 @@ ScrollView {
+                    // Prevent loop.
+                    if (shouldUpdateClipboardText) {
+                        shouldUpdateClipboardText = false;
+                        // Remove any blank lines of macro expansions.
+                        var alteredText = clipboard.text();
+                        var selectionStartLine = TextUtilities.getLineNumberForPosition(alteredText, textArea.selectionStart);
+                        for (var macroIndex = 0; macroIndex < textArea.macros.length; ++macroIndex) {
+                            var macro = textArea.macros[macroIndex];
+                            if (macro["collapsed"] === false) {
+                                // Calculate where the blank lines start and end, and remove the text inbetween.
+                                var blankStart = TextUtilities.getLineStartForLine(alteredText, (macro["startLine"]+1-selectionStartLine));
+                                var blankEnd = TextUtilities.getLineStartForLine(alteredText, (macro["startLine"]+1-selectionStartLine+macro["lineCount"]));
+                                alteredText = [alteredText.slice(0, blankStart), alteredText.slice(blankEnd)].join('')
+                            }
+                        }
+                        clipboard.setText(alteredText);
+                        shouldUpdateClipboardText = true;
+                    }
+                }
             }
 
+
             //area left of the TextEdit, contains lineNumbers, errorMessages, Breakpoints,...
             Rectangle {
                 id: sidebar
                 focus: false
                 x: container.contentX/scale.zoom
                 y: 0
                 height: Math.max(container.height, textArea.contentHeight)
+                width: errorBar.width + fontMetrics.averageCharacterWidth + (fontMetrics.averageCharacterWidth * textArea.lineCount.toString().length);
                 color: "#eeeeeb"
-                width: {
-                    errorBar.width + fontMetrics.averageCharacterWidth + (fontMetrics.averageCharacterWidth * textArea.lineCount.toString().length);
+
+                property alias _errorBar: errorBar
+
+                function updateLineNumbers() {
+                    lineNumbersBar.updateLineNumbers();
                 }
 
                 //linenumbers
                 Column {
-                    id: linenumbers
-                    x: 1 + errorBar.width
+                    id: lineNumbersBar
+                    anchors.left: parent.left
+                    anchors.leftMargin: 3
                     y: textArea.textMargin/2
 
-                    Repeater {
-                        model: textArea.lineCount
-                        delegate: Text {
-                            color: "gray"
-                            font: textArea.font
-                            text: index + 1
-                            height: fontMetrics.height
+                    property var _lineNumberObjects: []

```

Comment:
I think you could do this way easier:
Take the old implementation and set `text: lineNumberLabels[index]`. Then create a `updateLineNumbers()` signal in qml, which indicates to the text delegates that the lineNumberLabels array changed. In the text delegates, you add a `Connections` elements which updates the text in `onUpdateLineNumbers: text = lineNumberLabels[index]`. You can also try to only do a full update of the line number objects/labels if there really are macros.
That might work and would eliminate the need to recreate all line number objects all the time.

The thing that I'm not really happy about in the current implementation is that you always have to do this potentially expensive line number updating, even if there are no macros and there is no need for this.